ee/desktop/notify/notify_linux.go

```diff
@@ -4,38 +4,149 @@
 package notify
 
 import (
+	"context"
 	"fmt"
+	"net/url"
 	"os/exec"
+	"sync"
+	"time"
 
+	"github.com/go-kit/kit/log"
 	"github.com/go-kit/kit/log/level"
 	"github.com/godbus/dbus/v5"
 )
 
-func (d *DesktopNotifier) SendNotification(title, body string) error {
-	if err := d.sendNotificationViaDbus(title, body); err == nil {
+type dbusNotifier struct {
+	iconFilepath        string
+	logger              log.Logger
+	conn                *dbus.Conn
+	signal              chan *dbus.Signal
+	interrupt           chan struct{}
+	sentNotificationIds map[uint32]bool
+	lock                sync.RWMutex
+}
+
+const (
+	notificationServiceObj       = "/org/freedesktop/Notifications"
+	notificationServiceInterface = "org.freedesktop.Notifications"
+	signalActionInvoked          = "org.freedesktop.Notifications.ActionInvoked"
+)
+
+func newOsSpecificNotifier(logger log.Logger, iconFilepath string) *dbusNotifier {
+	conn, err := dbus.ConnectSessionBus()
+	if err != nil {
+		level.Warn(logger).Log("msg", "couldn't connect to dbus to start notifier listener, proceeding without it", "err", err)
+	}
+
+	return &dbusNotifier{
+		iconFilepath:        iconFilepath,
+		logger:              logger,
+		conn:                conn,
+		signal:              make(chan *dbus.Signal),
+		interrupt:           make(chan struct{}),
+		sentNotificationIds: make(map[uint32]bool),
+		lock:                sync.RWMutex{},
+	}
+}
+
+func (d *dbusNotifier) Listen() error {
+	if d.conn != nil {
+		if err := d.conn.AddMatchSignal(
+			dbus.WithMatchObjectPath(notificationServiceObj),
+			dbus.WithMatchInterface(notificationServiceInterface),
+		); err != nil {
+			level.Error(d.logger).Log("msg", "couldn't add match signal", "err", err)
+			return fmt.Errorf("couldn't register to listen to signals in dbus: %w", err)
+		}
+		d.conn.Signal(d.signal)
+	} else {
+		level.Warn(d.logger).Log("msg", "cannot set up DBUS listener -- no connection to session bus")
+	}
+
+	for {
+		select {
+		case signal := <-d.signal:
+			if signal == nil || signal.Name != signalActionInvoked {
+				continue
+			}
+
+			// Confirm that this is a Kolide-originated notification by checking for known notification IDs
+			notificationId := signal.Body[0].(uint32)
+			d.lock.RLock()
+			if _, found := d.sentNotificationIds[notificationId]; !found {
+				// This notification didn't come from us -- ignore it
+				d.lock.RUnlock()
+				continue
+			}
+			d.lock.RUnlock()
+
+			// Confirm that the URI is legitimate
+			actionUri := signal.Body[1].(string)
+			_, err := url.Parse(actionUri)
+			if err != nil {
+				level.Error(d.logger).Log("msg", "received invalid action URI from dbus", "action_uri", actionUri, "parse_err", err)
+				continue
+			}
```

Comment:
we do the same check in notify_consumer.go, can we drop it here?

Reply:
Ah yup, forgot to pull it out here! Will update